src/providers/trusted_service/context/key_management.rs

```diff
@@ -0,0 +1,146 @@
+use psa_crypto::types::status::{Error, Status};
+use std::convert::{TryFrom, TryInto};
+use std::mem::forget;
+use std::ptr::null_mut;
+
+impl Context {
+    pub fn generate_key(
+        &self,
+        key_attrs: Attributes,
+        id: u32,
+    ) -> parsec_interface::requests::Result<u32> {
+        let proto_req = Box::new(GenerateKeyIn {
+            attributes: Some(KeyAttributes {
+                r#type: u16::try_from(key_attrs.key_type)? as u32,
+                key_bits: key_attrs.bits.try_into()?,
+                lifetime: KeyLifetime::Persistent as u32,
+                id,
+                policy: Some(KeyPolicy {
+                    usage: key_attrs.policy.usage_flags.try_into()?,
+                    alg: key_attrs.policy.permitted_algorithms.try_into()?,
+                }),
+            }),
+        });
+        let mut proto_resp = Box::new(GenerateKeyOut::default());
+        Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::GenerateKey,
+            self.rpc_caller,
+        )?;
+
+        Ok(0)
+    }
+
+    pub fn destroy_key(&self, id: u32) -> parsec_interface::requests::Result<()> {
+        if !self.check_key_exists(id)? {
+            return Err(ResponseStatus::PsaErrorDoesNotExist);
+        }
+        let proto_req = Box::new(OpenKeyIn { id });
+        let mut proto_resp = Box::new(OpenKeyOut::default());
+        Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::OpenKey,
+            self.rpc_caller,
+        )?;
+        let handle = proto_resp.handle;
+
+        let proto_req = Box::new(DestroyKeyIn { handle });
+        let mut proto_resp = Box::new(DestroyKeyOut::default());
+        Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::DestroyKey,
+            self.rpc_caller,
+        )?;
+        Ok(())
+    }
+
+    pub fn check_key_exists(&self, id: u32) -> Result<bool, Error> {
+        let proto_req = Box::new(OpenKeyIn { id });
+        let mut proto_resp = Box::new(OpenKeyOut::default());
+        match Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::OpenKey,
+            self.rpc_caller,
+        ) {
+            Ok(_) => {
+                let handle = proto_resp.handle;
+                let proto_req = Box::new(CloseKeyIn { handle });
+                Context::send_request(
+                    &proto_req,
+                    &mut proto_resp,
+                    Opcode::OpenKey,
+                    self.rpc_caller,
+                )?;
+                Ok(true)
+            }
+            Err(e) => {
+                if e == Error::DoesNotExist {
+                    Ok(false)
+                } else {
+                    Err(e)
+                }
+            }
+        }
+    }
+
+    fn send_request<Req: Message, Resp: Message>(
+        req: &Box<Req>,
+        resp: &mut Box<Resp>,
```

Comment:
is that the same as:
```
    fn send_request(
        req: &Box<Message>,
        resp: &mut Box<Message>,
```

Reply:
and also maybe same as:
```
    fn send_request(
        req: &dyn Message,
        resp: &mut dyn Message,
```
why might be easier to read?

Reply:
The second one doesn't work because `dyn Message` isn't `Sized` so it won't compile.
The first one would look like this:
```rust
    fn send_request(
        req: &Box<dyn Message>,
        resp: &mut Box<dyn Message>,
```

But I'd then have to create the messages like this: `let proto_req: Box<dyn Message> = Box::from(CloseKeyIn { handle });` which means I can't access the fields inside anymore, only the methods defined on `Message`. I think the approach with `<Resp: Message, Req: Message>` just says "any type that implements `Message`", instead of "this particular type, `Box<dyn Message>`".

Reply:
Ah, the 2nd one works with `impl` instead of `dyn`!